src/kv/experimental.h

```diff
@@ -0,0 +1,341 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ds/hash.h"
+#include "kv_types.h"
+#include "map.h"
+#include "tx_view.h"
+
+#include <vector>
+
+namespace kv
+{
+  namespace experimental
+  {
+    using SerialisedRep = std::vector<uint8_t>;
+
+    // TODO: I don't think this needs to be customisable? If the map is storing
+    // _your_ types as keys, you might need to tell it how to compare them. But
+    // we know how to compare byte-vectors, and this is an internal detail so
+    // why would you change it?
+    using RepHasher = std::hash<SerialisedRep>;
```

Comment:
I'm not sure this is right, because for T t1, t2 where t1 == t2, it definitely seems possible that serialised(t1) != serialised(t2) (eg. "10.0" and "1e1"), as well as the opposite (NaN).

Reply:
For posterity, a brief summary of what's been discussed offline:

We agree the _hashing_ is an irrelevant detail we can hide from users, but we need to be precise with the equality implications of pluggable serialisers. Essentially we'd like injective serialisers, and this injectivity should be maintained across versions (ie - if you previously emit `{"foo": 1, "bar": 2}`, you shouldn't in future emit `{"bar": 2, "foo": 1}`).

For our serialisers, we'd like to eventually use verified code with a formal proof that it meets this property. For user-defined serialisers, we need to be clear about this in documentation: equality and storage deal only over the serialised forms, and these typed wrappers are just a convenience.